Document create/drop catalog (#19612)
trinodb/trino: docs/src/main/sphinx/admin/properties-catalog.md

# Catalog management properties

(prop-catalog-management)=

## `catalog.management`

- **Type:** {ref}`prop-type-string`
- **Allowed values:** `static`, `dynamic`
- **Default value:** `static`

Trino reads catalog property files and configures available catalogs on server
startup. When set to `dynamic`, catalog configuration can also be managed using
[](/sql/create-catalog) and [](/sql/drop-catalog). Creating and dropping

> catalogs using these SQL commands will create new and delete existing catalog

Author: I guess this is only true if `catalog.store=file`, which is the default, but since it's not documented, maybe we can update this when documenting `catalog.store` and other properties?

Reviewer: We should document these properties now as well

Reviewer: ```suggestion
catalogs using these SQL commands adds and removes catalog
```

property files.

:::{warning}
This feature is experimental only. Because of the security implications the
syntax might change and be backward incompatible.
:::

:::{warning}
Some connectors are known not to release all resources when dropping a catalog
that uses such connector. This includes all connectors that can read data from

> HDFS, which are {doc}`/connector/hive`, {doc}`/connector/iceberg`,

Reviewer 2: I'd mention S3, GCS, and Azure

{doc}`/connector/delta-lake`, and {doc}`/connector/hudi`.
:::

:::{warning}
The complete CREATE CATALOG query can be logged, and visible in the {doc}`Web
UI </admin/web-interface>`. This includes any sensitive properties, like
passwords and other credentials. See {doc}`secrets </security/secrets>`.
:::

> ## `catalog.prune.update-interval`

Author: I just noticed I missed `catalog.prune.enabled` but I'm not sure what's the use case for disabling it.

Reviewer 2: I'm not sure why I added that.  Maybe it made testing easier.  Anyway, I'm not sure we need to document the enable

# Catalog management properties

(prop-catalog-management)=

> 

Reviewer: no empty line needed between title and anchor

## `catalog.management`

- **Type:** {ref}`prop-type-string`
- **Allowed values:** `static`, `dynamic`
- **Default value:** `static`

Trino reads catalog property files and configures available catalogs on server
startup. When set to `dynamic`, catalog configuration can also be managed using
[](/sql/create-catalog) and [](/sql/drop-catalog). Creating and dropping
catalogs using these SQL commands will create new and delete existing catalog

> property files on all nodes.

Reviewer: We also need to clarify that it requires read access to the filesystem on the coordinator and workers by the process running Trino.

Reviewer: Also .. what happens if a new node is added via cluster scaling for example on k8s .. will it somehow get the catalog files created as necessary or not?

Author: I tested this locally in containers and workers get all the catalogs when joining the cluster (they're able to execute queries), but they don't write the files to disk. I updated relevant parts of the docs.

:::{warning}

> This feature is experimental only. Because of the security implications the

Reviewer: I dont think the syntax changes and the security implications are necessarily related.

Author: I'd leave this, we can always remove this later when we figure out how to handle sensitive values even if there are no syntax changes.

Reviewer: ok

syntax might change and be backward incompatible.
:::

> :::{warning}

Reviewer: Do we really still need this. I thought this was a problem from the Hadoop dependencies and those are all gone. And even if .. its only applicable for LOTS of create/drop cycles .. that should be rare.

Author: I was testing this a few days ago and there was a 10% increase in non-heap memory usage after creating and dropping a hive catalog 15 times.

Reviewer: Lets leave it for now .. but we have to keep an eye on it and remove it when it no longer applies

Some connectors are known not to release all resources when dropping a catalog
that uses such connector. This includes all connectors that can read data from
HDFS, S3, GCS, or Azure, which are {doc}`/connector/hive`,
{doc}`/connector/iceberg`, {doc}`/connector/delta-lake`, and
{doc}`/connector/hudi`.
:::

:::{warning}

> The complete CREATE CATALOG query can be logged, and visible in the {doc}`Web

Reviewer: ```suggestion
The complete `CREATE CATALOG` query is logged, and visible in the {doc}`Web
```

UI </admin/web-interface>`. This includes any sensitive properties, like
passwords and other credentials. See {doc}`secrets </security/secrets>`.
:::

## `catalog.prune.update-interval`

- **Type:** {ref}`prop-type-duration`
- **Default value:** `5s`
- **Minimum value:** `1s`

> Interval for pruning dropped catalogs. Dropping a catalog does not interrupt

Reviewer: we should add info on where they are dropped .. on coordinator and all workers..

Author: This is not relevant here, but I explained that in the description of other properties.

# Catalog management properties

(prop-catalog-management)=
## `catalog.management`

> - **Type:** {ref}`prop-type-string`

Reviewer: All these links that use 

```
{ref}`prop-type-string`
```

should be changed to use the shorter markdown syntax

```
[](prop-type-string)
```

- **Allowed values:** `static`, `dynamic`
- **Default value:** `static`

Trino reads catalog property files and configures available catalogs on server
startup. When set to `dynamic`, catalog configuration can also be managed using
[](/sql/create-catalog) and [](/sql/drop-catalog).

:::{warning}
This feature is experimental only. Because of the security implications the
syntax might change and be backward incompatible.
:::

:::{warning}
Some connectors are known not to release all resources when dropping a catalog
that uses such connector. This includes all connectors that can read data from

> HDFS, S3, GCS, or Azure, which are {doc}`/connector/hive`,

Reviewer: All these links should change in format to 

```
[](/connector/hive)
```

{doc}`/connector/iceberg`, {doc}`/connector/delta-lake`, and
{doc}`/connector/hudi`.
:::

:::{warning}
The complete `CREATE CATALOG` query is logged, and visible in the {doc}`Web UI

> </admin/web-interface>`. This includes any sensitive properties, like passwords

Reviewer: change link syntax

Requires [](prop-catalog-management) to be set to `dynamic`. When set to
`file`, creating and dropping catalogs using the SQL commands adds and removes
catalog property files on the coordinator node. Trino server process requires
write access in the catalog configuration directory. Existing catalog files are
also read on the coordinator startup. When set to `memory`, catalog
configuration is only managed in memory, and any existing files are ignored on
startup.

## `catalog.config-dir`

- **Type:** [](prop-type-string)
- **Default value:** `etc/catalog/`

Requires [](prop-catalog-management) to be set to `static` or
[](prop-catalog-store) to be set to `file`. The directory with catalog property files.

## `catalog.disabled-catalogs`

- **Type:** [](prop-type-string)

Requires [](prop-catalog-management) to be set to `static` or
[](prop-catalog-store) to be set to `file`. Comma-separated list of catalogs
to ignore on startup.

## `catalog.read-only`

- **Type:** [](prop-type-string)
- **Default value:** `false`

Requires [](prop-catalog-store) to be set to `file`. If true, existing
catalog property files cannot be deleted or modified and new catalog property

> files cannot be created.

Reviewer 3: how is this achieved? I don't see this happenning on the filesystem level. Maybe we mean to say that enabling this disables `CREATE CATALOG` and `DROP CATALOG`.

Additionally with catalog store set to file and mode set to dynamic Trino won't (AFAIK) "reload" changed files or newly added file to catalog directory by hand - this needs to be called out IMO.

Author: https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/connector/FileCatalogStore.java#L131

Author: Why it needs to be called out? There has been no reload mechanism in Trino even with the pre-existing static catalog management.

Reviewer 3: > https://github.com/trinodb/trino/blob/master/core/trino-main/src/main/java/io/trino/connector/FileCatalogStore.java#L131

Yes, so  it disables `CREATE` and `DROP` only - it doesn't disable "existing catalog property files cannot be deleted or modified and new catalog property files cannot be created".

Though I'm interested about when `dynamic` + `read-only` makes sense. It's probably a question to @dain.

Reviewer: Lets reword this accordingly

Author: I don't understand the difference. @mosabua you can push a change directly to this branch.

Reviewer: Done

- **Allowed values:** `file`, `memory`
- **Default value:** `file`

Requires [](prop-catalog-management) to be set to `dynamic`. When set to
`file`, creating and dropping catalogs using the SQL commands adds and removes
catalog property files on the coordinator node. Trino server process requires
write access in the catalog configuration directory. Existing catalog files are
also read on the coordinator startup. When set to `memory`, catalog
configuration is only managed in memory, and any existing files are ignored on
startup.

## `catalog.config-dir`

- **Type:** [](prop-type-string)
- **Default value:** `etc/catalog/`

Requires [](prop-catalog-management) to be set to `static` or
[](prop-catalog-store) to be set to `file`. The directory with catalog property
files.

## `catalog.disabled-catalogs`

- **Type:** [](prop-type-string)

Requires [](prop-catalog-management) to be set to `static` or
[](prop-catalog-store) to be set to `file`. Comma-separated list of catalogs to
ignore on startup.

## `catalog.read-only`

- **Type:** [](prop-type-string)
- **Default value:** `false`

Requires [](prop-catalog-store) to be set to `file`. If true, existing catalog

> property files cannot be removed with `DROP CATALOG`, and now new catalog files

Author: There's a typo `s/now/no`

Reviewer: Dang.. can you fix that in your follow up PR maybe?
